[SPARK-19507][PySpark][SQL] Show field name in _verify_type error

dgingrich opened the pull request:
## What changes were proposed in this pull request?

Add better error messages to _verify_type to track down which columns are not compliant with the schema.

## How was this patch tested?

Unit tests, doctest, hand inspection in REPL.

## Legal

This is my original work and I license the work to the project under the project’s open source license.

Commit:
Remove "# noqa" comment from docstring

Commit:
WIP: Add name parameter and better debugging to _verify_types

* Add name paramter to _verify_types
* Include name parameter in debug messages
* Build name message for nested structs, arrays, and maps
* Add detailed tests to flesh out spec for _verify_types (WIP)

dgingrich commented:
Opening as a comparison for https://github.com/apache/spark/pull/17213

AmplabJenkins commented:
Can one of the admins verify this patch?

holdenk commented:
Jenkins, ok to test.

SparkQA commented:
**[Test build #74284 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74284/testReport)** for PR 17227 at commit [`5f72a54`](https://github.com/apache/spark/commit/5f72a547a948b5c5a787aace52df04bc8503888b).

holdenk commented:
I'm a bit busy today but I'll try and take a look through this and https://github.com/apache/spark/pull/17213 sometime this week. If y'all agree on which one you think is better though I can focus on reviewing that.

SparkQA commented:
**[Test build #74284 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74284/testReport)** for PR 17227 at commit [`5f72a54`](https://github.com/apache/spark/commit/5f72a547a948b5c5a787aace52df04bc8503888b).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74284/
Test FAILed.

HyukjinKwon commented:
Yea, if the change is not too different, I would rather help the review the first PR before taking over.

HyukjinKwon commented:
BTW, let's put `[WIP]` in the title if it is.

HyukjinKwon commented:
(Let's proceed with this one per https://github.com/apache/spark/pull/17213#issuecomment-285530248)

blrnw3 commented:
👍

Commit:
Finish types._verify_type tests.

dgingrich commented:
Tests finished, ready for review.

SparkQA commented:
**[Test build #74430 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74430/testReport)** for PR 17227 at commit [`5c5ab5b`](https://github.com/apache/spark/commit/5c5ab5b86b0995d8b955755ab972f8c5f1474956).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74430/
Test FAILed.

dgingrich commented:
Jenkins test failures look unrelated to my changes (I was actually seeing them locally w/ `run-tests` so I tested my changes in isolation w/ `pytest` & some hacks).  What's the criteria for accepting the patch?  If the build has to be clean I'll rebase to master and see if the errors clear up.

dgingrich commented:
@blrnw3 if you want to code review (though things might change, re: Jenkins failures)

blrnw3 commented:
From my perspective you've covered everything, so thanks for pushing this through.

SparkQA commented:
**[Test build #74548 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74548/testReport)** for PR 17227 at commit [`fcd2067`](https://github.com/apache/spark/commit/fcd2067c97e60a72653cd34905727e827c95ccbd).

SparkQA commented:
**[Test build #74548 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74548/testReport)** for PR 17227 at commit [`fcd2067`](https://github.com/apache/spark/commit/fcd2067c97e60a72653cd34905727e827c95ccbd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `        # obj, data_type, exception (None for success or Exception subclass for error)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74548/
Test PASSed.

dgingrich commented:
Looks like the rebase fixed the tests.  This is good to go from my POV, let me know if you need any changes.

dgingrich commented:
Ping!  Let me know if you need more work from me.

gatorsmile commented:
cc @ueshin

ueshin commented on python/pyspark/sql/tests.py:
```diff
@@ -2367,6 +2380,151 @@ def range_frame_match():
+                (None, StructType([]))]:
+            _verify_type(obj, data_type, nullable=True)
+            msg = "_verify_type(%s, %s, nullable=True)" % (obj, data_type)
+            self.assertTrue(True, msg)
```
I think we should surround `_verify_type(obj, data_type, nullable=True)` with try block and check if it raises an exception or not as the same as we do in `test_verify_type_not_nullable` test.

dgingrich replied:
Added

ueshin reviewed (commented):
LGTM, except for some comments.

dgingrich commented:
Thanks for the comments @ueshin, will address them this weekend.

dgingrich commented on python/pyspark/sql/tests.py:
```diff
@@ -2367,6 +2380,151 @@ def range_frame_match():
+
+            # Array
+            ([], ArrayType(IntegerType()), None),
+            (["1", None], ArrayType(StringType(), containsNull=True), None),
```
Added

dgingrich commented on python/pyspark/sql/tests.py:
```diff
@@ -2367,6 +2380,151 @@ def range_frame_match():
+            ({"a": 1}, MapType(StringType(), IntegerType()), None),
+            ({"a": 1}, MapType(IntegerType(), IntegerType()), TypeError),
+            ({"a": "1"}, MapType(StringType(), IntegerType()), TypeError),
+            ({"a": None}, MapType(StringType(), IntegerType(), valueContainsNull=True), None),
```
Added

dgingrich commented on python/pyspark/sql/tests.py:
```diff
@@ -2367,6 +2380,151 @@ def range_frame_match():
+            (MyObj(s="a", i=1), MyStructType, None),
+            (MyObj(s="a", i=None), MyStructType, None),
+            (MyObj(s="a"), MyStructType, None),
+            (MyObj(s="a", i="1"), MyStructType, TypeError),
```
Added

dgingrich commented on python/pyspark/sql/tests.py:
```diff
@@ -2367,6 +2380,151 @@ def range_frame_match():
+                except Exception as e:
+                    traceback.print_exc()
+                    self.fail(msg)
+                self.assertTrue(True, msg)
```
IIRC that was required at some point, I think to get a test runner to pick up the test.  But I just tried removing it and tests ran with pytest and the Python3 unittest runner, so removed the line here and in `test_verify_type_ok_nullable`.  We should probably look at the full test suite output from Jenkins to make sure the tests are run under Python2.

Commit:
Cleanup (PR feedback)

* Change "test_verify_type_ok_nullable" to match "test_verify_type_not_nullable"
* Add more null value tests
* Remove unused "self.assertTrue(true)" lines

dgingrich commented:
Pushed new commit addressing PR feedback.  I had to create a new local environment (new laptop) and was getting test errors on unrelated code before my changes.  I'm hoping that's just my laptop and Jenkins will pass.  If not I'll retry rebasing.

SparkQA commented:
**[Test build #78324 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78324/testReport)** for PR 17227 at commit [`5b4324e`](https://github.com/apache/spark/commit/5b4324e966a71fff787625b60001566f2895ac6a).

SparkQA commented:
**[Test build #78324 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78324/testReport)** for PR 17227 at commit [`5b4324e`](https://github.com/apache/spark/commit/5b4324e966a71fff787625b60001566f2895ac6a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78324/
Test FAILed.

shaneknapp commented:
test this please

SparkQA commented:
**[Test build #78329 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78329/testReport)** for PR 17227 at commit [`5b4324e`](https://github.com/apache/spark/commit/5b4324e966a71fff787625b60001566f2895ac6a).

SparkQA commented:
**[Test build #78329 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78329/testReport)** for PR 17227 at commit [`5b4324e`](https://github.com/apache/spark/commit/5b4324e966a71fff787625b60001566f2895ac6a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78329/
Test PASSed.

ueshin commented:
LGTM, do you have any other concerns? @holdenk, @HyukjinKwon

HyukjinKwon commented on python/pyspark/sql/tests.py:
```diff
@@ -30,6 +30,19 @@
 import datetime
+import traceback
+
+if sys.version_info[:2] <= (2, 6):
```
Not a big deal but I guess we dropped 2.6 support.

dgingrich replied:
Looks like most of the other tests still have the `<= (2, 6)` check (see python/pyspark/ml/tests.py) so leaving in place.

HyukjinKwon replied:
Yea, let's leave it then. Not a big deal.

HyukjinKwon commented on python/pyspark/sql/tests.py:
```diff
@@ -2367,6 +2380,157 @@ def range_frame_match():
+                (None, IntegerType()),
+                (None, FloatType()),
+                (None, StringType()),
+                (None, StructType([]))]:
```
Not a big deal too. Could we just take this out in a separate variable?

dgingrich replied:
Meaning remove the None from the tuples and just loop over the Types?  I like it a little better as is since the tuples are basically `_verify_type`'s args but am fine with either.  Let me know which you prefer and I can change the code.

HyukjinKwon replied:
Let's do this like ...

```python
types = [IntegerType()), FloatType()), StringType()), StructType([])]
for ...
```

if you don't mind. I think taking out the same value in a loop is slightly better.

dgingrich replied:
Sure, either works for me.  Changed.

HyukjinKwon commented on python/pyspark/rdd.py:
```diff
@@ -627,7 +627,6 @@ def sortPartition(iterator):
     def sortByKey(self, ascending=True, numPartitions=None, keyfunc=lambda x: x):
         """
         Sorts this RDD, which is assumed to consist of (key, value) pairs.
-        # noqa
```
(I have no idea why this was added in the first place ...)

HyukjinKwon commented on python/pyspark/sql/types.py:
```diff
@@ -1300,70 +1300,80 @@ def _verify_type(obj, dataType, nullable=True):
             return
         else:
-            raise ValueError("This field is not nullable, but got None")
+            raise ValueError("%s: This field is not nullable, but got None" % name)
```
Probably, I missed something. However, is there any test case that actually checks this message change?

dgingrich replied:
No, I never check the actual exception message.  I normally don't check the contents of exception messages since they shouldn't be used programmatically (the tests are mostly to exercise all code paths to make sure I didn't break something).

But it makes sense to test the prefix is set since that's the main point of the change.  Added a test looking at the exception message prefix, which should be robust.

HyukjinKwon commented on python/pyspark/sql/tests.py:
```diff
@@ -2367,6 +2380,157 @@ def range_frame_match():
+            if exp is None:
+                try:
+                    _verify_type(obj, data_type, nullable=False)
+                except Exception as e:
```
It looks this `e` is not used.

dgingrich replied:
Removed the `e`.

HyukjinKwon commented on python/pyspark/sql/types.py:
```diff
@@ -1249,7 +1249,7 @@ def _infer_schema_type(obj, dataType):
 
 
-def _verify_type(obj, dataType, nullable=True):
+def _verify_type(obj, dataType, nullable=True, name="obj"):
```
Just a question. @dgingrich Do you maybe know if there is any change that "obj" is printed instead? It is rather a nitpick but I would think it is odds if it prints "obj".

dgingrich replied:
This will print "obj" when called from `session.createDataFrame` (https://github.com/dgingrich/spark/blob/topic-spark-19507-verify-types/python/pyspark/sql/session.py#L408).  It'd be easy to set the name where it's called but it wasn't clear what to set it to.   The input can be either an RDD, list, or pandas.DataFrame.

HyukjinKwon replied:
I meant this case:

```python
>>> from pyspark.sql.types import *
>>> spark.createDataFrame(["a"], StringType()).printSchema()
```

```
root
 |-- value: string (nullable = true)
```
```python
>>> from pyspark.sql.types import *
>>> spark.createDataFrame(["a"], IntegerType()).printSchema()
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../spark/python/pyspark/sql/session.py", line 526, in createDataFrame
    rdd, schema = self._createFromLocal(map(prepare, data), schema)
  File ".../spark/python/pyspark/sql/session.py", line 387, in _createFromLocal
    data = list(data)
  File ".../spark/python/pyspark/sql/session.py", line 516, in prepare
    verify_func(obj, dataType)
  File ".../spark/python/pyspark/sql/types.py", line 1326, in _verify_type
    % (name, dataType, obj, type(obj)))
TypeError: obj: IntegerType can not accept object 'a' in type <type 'str'>
```

It sounds "obj" should be "value". It looks we should specify the name around https://github.com/dgingrich/spark/blob/topic-spark-19507-verify-types/python/pyspark/sql/session.py#L516.

HyukjinKwon replied:
Let's fix this case.

```python
>>> from pyspark.sql.types import *
>>> spark.createDataFrame(["a"], StringType()).printSchema()
```

```
root
 |-- value: string (nullable = true)
```
```python
>>> from pyspark.sql.types import *
>>> spark.createDataFrame(["a"], IntegerType()).printSchema()
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../spark/python/pyspark/sql/session.py", line 526, in createDataFrame
    rdd, schema = self._createFromLocal(map(prepare, data), schema)
  File ".../spark/python/pyspark/sql/session.py", line 387, in _createFromLocal
    data = list(data)
  File ".../spark/python/pyspark/sql/session.py", line 516, in prepare
    verify_func(obj, dataType)
  File ".../spark/python/pyspark/sql/types.py", line 1326, in _verify_type
    % (name, dataType, obj, type(obj)))
TypeError: obj: IntegerType can not accept object 'a' in type <type 'str'>
```

It sounds "obj" should be "value". It looks we should specify the name around https://github.com/dgingrich/spark/blob/topic-spark-19507-verify-types/python/pyspark/sql/session.py#L516.

HyukjinKwon replied:
I guess this is only place that we print "obj" maybe? If so, let's set `name=None`.

dgingrich replied:
Set `name=value` in the call at session.py line 516.  

It will still print `obj` if the schema is a StructType: `TypeError: obj.a: MyStructType can not accept object 'a' in type <type 'str'>`.  Would you like to change that too?

Right now changing the default name to None would make the error message worse: `TypeError: None: IntegerType can not accept object 'a' in type <type 'str'>`.  

The best way to make the error message pretty is probably:
- Set the default name to None
- If name==None, don't prepend the `%s: ` string to the error messages

That would make your exmple: `TypeError: IntegerType can not accept object 'a' in type <type 'str'>`.  

IMO `obj` is not as pretty but reasonable since it's so simple.  Let me know what you prefer.  My only goal is that next time I get a schema failure it tells me what field to look at :)

HyukjinKwon reviewed (commented):
Thanks for pinging me @ueshin. I nitpicked some. Otherwise, looks pretty good to me. @holdenk's sing-off might be nicer.

Commit:
Cleanup (PR feedback)

* Test that _verify_type exceptions start with passed name
* Remove unused Exception variable

SparkQA commented:
**[Test build #78472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78472/testReport)** for PR 17227 at commit [`2351153`](https://github.com/apache/spark/commit/23511537966577bd6d2b4bbe9dd898faa0e72e97).

SparkQA commented:
**[Test build #78472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78472/testReport)** for PR 17227 at commit [`2351153`](https://github.com/apache/spark/commit/23511537966577bd6d2b4bbe9dd898faa0e72e97).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78472/
Test PASSed.

Commit:
Cleanup (PR feedback)

* Set name=value when calling _verify_type from createDataFrame for a DataType
* Remove Nones from testing tuples in test_verify_type_ok_nullable

SparkQA commented:
**[Test build #78657 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78657/testReport)** for PR 17227 at commit [`566f759`](https://github.com/apache/spark/commit/566f759aca61d4f44628bae0265fd930bff386ed).

SparkQA commented:
**[Test build #78657 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78657/testReport)** for PR 17227 at commit [`566f759`](https://github.com/apache/spark/commit/566f759aca61d4f44628bae0265fd930bff386ed).
 * This patch **fails PySpark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78657/
Test FAILed.

dgingrich commented:
Oops, fixed missing quotes bug in `session.py`.

SparkQA commented:
**[Test build #78740 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78740/testReport)** for PR 17227 at commit [`6c1e0b6`](https://github.com/apache/spark/commit/6c1e0b690bdd1914b5056c8b2934614534c622cb).

SparkQA commented:
**[Test build #78740 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78740/testReport)** for PR 17227 at commit [`6c1e0b6`](https://github.com/apache/spark/commit/6c1e0b690bdd1914b5056c8b2934614534c622cb).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78740/
Test PASSed.

ueshin commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #79058 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79058/testReport)** for PR 17227 at commit [`6c1e0b6`](https://github.com/apache/spark/commit/6c1e0b690bdd1914b5056c8b2934614534c622cb).

SparkQA commented:
**[Test build #79058 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79058/testReport)** for PR 17227 at commit [`6c1e0b6`](https://github.com/apache/spark/commit/6c1e0b690bdd1914b5056c8b2934614534c622cb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79058/
Test PASSed.

HyukjinKwon reviewed (commented):
Others look good to me but I think https://github.com/apache/spark/pull/17227#discussion_r124135474 should be fixed. Otherwise, it will print out as below, as @dgingrich pointed out:

```
from pyspark.sql.types import *
schema = StructType([StructField('s', IntegerType(), nullable=False)])
spark.createDataFrame([["1"]], schema)
...
TypeError: obj.s: IntegerType can not accept object '1' in type <type 'str'>
```

I think we don't print as such anywhere. I don't want to block this PR. Let me prepare a PR to your branch. Please review the change.

HyukjinKwon commented:
@dgingrich, I tried to address my comments at my best here - https://github.com/dgingrich/spark/pull/1. Could you review that change and merge it if it looks good to you so that the change is merged into this PR?

HyukjinKwon commented:
Wait ... this is possibly a hot path that requires row-per-operation ... @ueshin and @dgingrich, I think we should re-write this to avoid record-per-type dispatch ... For me, I wouldn't mind if we go and merge my approach, dgingrich#1 here for now. I will make a follow up right after it gets merged.
